cmd/shadowsocks-local/local.go

```diff
@@ -1,430 +1,814 @@
-		if servers.failCnt[i] > 0 && rand.Intn(servers.failCnt[i]+baseFailCnt) != 0 {
-			skipped = append(skipped, i)
-			continue
-		}
-		remote, err = connectToServer(i, rawaddr, addr)
-		if err == nil {
-			return
-		}
+// dial the server establish the ss connection
+func connectToServerUDP(ciph encrypt.Cipher, timeout int) (net.PacketConn, error) {
+	if ciph == nil {
+		return nil, ErrNilCipher
 	}
-	// last resort, try skipped servers, not likely to succeed
-	for _, i := range skipped {
-		remote, err = connectToServer(i, rawaddr, addr)
-		if err == nil {
-			return
-		}
+
+	sconn, err := ss.SecureListenPacket("udp", "", ciph.Copy(), timeout)
+	if err != nil {
+		return nil, err
 	}
-	return nil, err
+	ss.Logger.Debug("ss local connecting to server with UDP", zap.Int("timeout", timeout))
+	return sconn, nil
 }
 
-func handleConnection(conn net.Conn) {
-	if debug {
-		debug.Printf("socks connect from %s\n", conn.RemoteAddr().String())
+// handle the local socks5 connection and request remote server
+func handleConnection(server string, conn net.Conn, timeout int) {
+	ss.Logger.Debug("handle socks5 connect", zap.Stringer("source", conn.LocalAddr()), zap.Stringer("remote", conn.RemoteAddr()))
+	var err error
+
+	// After handshake ss will read the requset fron client to establish the proxy connection
+	// target is the socks5 request addr+port
+	target, err := getRequest(conn)
+	if err != nil {
+		ss.Logger.Error("error in getting socks5 request", zap.Error(err))
+		return
 	}
-	closed := false
-	defer func() {
-		if !closed {
-			conn.Close()
-		}
-	}()
 
-	var err error = nil
-	if err = handShake(conn); err != nil {
-		log.Println("socks handshake:", err)
+	// Sending connection established message immediately to client.
+	// This cost some round trip time for creating socks connection with the client.
+	// But if connection failed, the client will get connection reset error.
+	//
+	// Notice that the server response bind addr & port could be ignore by the socks5 client
+	// 0x00 0x00 0x00 0x00 0x00 0x00 is meaning less for bind addr block.
+	_, err = conn.Write([]byte{0x05, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00})
+	if err != nil {
+		ss.Logger.Error("send connection confirmation", zap.Error(err))
 		return
 	}
-	rawaddr, addr, err := getRequest(conn)
+
+	// after socks5 hand shake, we'd get a ss server connection
+	// cipher should be copied pass
+
+	ssconn, err := connectToServer(server, ciphers[server], timeout)
 	if err != nil {
-		log.Println("error getting request:", err)
+		ss.Logger.Error("erro in connect to ss server", zap.String("server", server), zap.Error(err))
 		return
 	}
-	// Sending connection established message immediately to client.
-	// This some round trip time for creating socks connection with the client.
-	// But if connection failed, the client will get connection reset error.
-	_, err = conn.Write([]byte{0x05, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x08, 0x43})
```

Comment:
I have been confused by this two magic number for a long time and read the rfc1928 again and again ==

Reply:
@dszhengyu Read the association and find how Socks5 handshaking and association works

Reply:
bnd.addr and bad.port will be ignored